### DISCUSSION TITLE
Create custom 8vim layouts easily using a spreadsheet

### DISCUSSION THREAD

**Discussion by kjoetom** [8VIM/8VIM#295] [Language Layout]
Hello everyone,

as there are occasional requests for modifications, I would like to offer an easy way for users to create their own Xpad layouts without much effort and using commonly available office software.

Attached you will find a spreadsheet that I created with LibreOffice Calc. You only need to enter the desired letters and characters for your layout, and then the content for a corresponding YAML file will be generated automatically. All you need to do is copy the generated content to an empty text file and save it with the .yaml extension.

You can then import this file into the settings of 8vim under "Select layout file from device".

The spreadsheet includes detailed instructions on how to proceed. Simply follow the steps outlined there to achieve the desired result.

The file was created with LibreOffice. The current version of the 8vim layout generator (2.x) no longer works in MS Excel. I have successfully tested it with a newer version of LibreOffice on both Linux and Windows.

You can download the file here: [8vim-layout-generator_LO-Calc_2.4.ods.zip](https://github.com/user-attachments/files/16826169/8vim-layout-generator_LO-Calc_2.4.ods.zip)

The current version 2.4 now also offers the possibility to integrate martin-brauns shortcuts (H4cker extension) into the layout (see [below](https://github.com/8VIM/8VIM/discussions/295#discussioncomment-6909222)). I have also added a hint on how to delete several letters at once. I also fixed a bug that caused 8vim to crash when loading the yaml file (see [below](https://github.com/8VIM/8VIM/discussions/295#discussioncomment-10443529)).

You can define not only individual letters, diacritical marks, and characters but also entire sentences, which can be quickly written using Xpad.

There are three ways to switch between uppercase and lowercase letters. Additionally, there is an optional feature to directly write numbers on the Xpad (see https://github.com/flide/8VIM/discussions/276).

The only thing stopping you now is yourself. 😉

Have fun!

kjoetom

That's what the spreadsheet looks like:

![layout](https://user-images.githubusercontent.com/77550697/230750370-112da1cf-f290-4e7c-8937-8d230f65b3f7.png)

**How to write the letters?**

To write the green letters, you should first write the corresponding base letter. Instead of going back into the circle, move your finger back to the previous sector (and potentially even further back) before finally returning to the circle to write the desired (green) letter.
Example for writing an "á"   ![green](https://user-images.githubusercontent.com/77550697/230750788-40d70428-6706-4820-9d53-4809b3814588.png)

To write the blue and red letters, you should also begin by writing the corresponding (black) base letter. Instead of going directly into the circle at the end, move your finger precisely one sector back and then forward one sector (and potentially even further forward), before returning to the circle to write the blue or desired red letter.
Example for writing a "ν"  (ny)   ![blue-red](https://user-images.githubusercontent.com/77550697/230750784-a05765b1-f413-44d3-9a59-e14ac71a73e2.png)

There are three ways to write uppercase letters:

1. By pressing the Shift button before writing a letter.  ![shift](https://user-images.githubusercontent.com/77550697/230800760-44e9445e-24ad-4058-bb4b-2428530db661.png)

2. By briefly moving your finger backward and forward before returning it to the center position to switch between upper and lower case for any letter (which in terms of continuous movement can also be realized as a circular motion).
Examples for writing a capital "A"

![capitalization-2a](https://user-images.githubusercontent.com/77550697/230750785-5cd9e92d-7ebe-4c19-85ba-267ef565bc08.png)  OR  ![capitalization-2b](https://user-images.githubusercontent.com/77550697/230750786-e3ebc05e-59fd-4472-86d3-8d9dc5425bfe.png)

3. By rotating your finger 360 degrees before returning it to the center position to switch between lowercase and uppercase for basic letters.
![capitalization-3](https://user-images.githubusercontent.com/77550697/230750787-b19e2c45-4597-47fe-8adb-cf3fa2fc612a.png)

**Comment by Glitchy-Tozier** [comment]
This looks amazing! With the layout optimizer and this, we have a great setup to quickly and easily develop and add layouts to the app! :)

A neat thing you could do is to upload the file (and its future updates) to a GitHub repo using the [`.fods` file format](https://superuser.com/questions/1264379/libreoffice-how-to-save-in-git-friendly-format#1264392), so the link in this thread always links to the newest version. Additionally it would open up the ability for others to contribute.

**Comment by kjoetom** [comment]
Thank you for the positive feedback. I really hope that these two tools will increase the usage of 8vim and perhaps even advance its further development.

I didn't follow your suggestion to upload a .fods file of the spreadsheet because I was unsure of the correct license that needed to be indicated (since Libreoffice code is used in the formulas). However, I don't see a big problem if a working version of the spreadsheet is simply available here and anyone who wants to can post it here again.

**Comment by martin-braun** [comment]
I like how you have incorporated the back and forth short-cut for capitalized letters, but can you import existing XMLs? Also can you access the numbers on the X-pad? What about other special characters that don't start in the center?

**Comment by kjoetom** [comment]
Unfortunately, it is not possible to import existing XMLs into the Calc or Excel spreadsheet. 
The letters, diacritics and signs you enter in the schematic layout all start from the centre. However, I have added the ability to enter numbers in the style of handwritten Arabic numbers. These all start in the left quadrant.

**Comment by kjoetom** [comment]
Custom layouts in xml format can no longer be loaded from 8vim iteration 12 onwards. This is due to a change from the xml format to the yaml format. Therefore I have updated the layout generator to version 2.x to be able to create the current custom layout format (yaml). But the previous xml version is still there in case someone wants or needs to use an older version of 8vim (iteration 11 or earlier).

kjoe

**Comment by kjoetom** [comment]
At the top of this post I have just uploaded a new version of the layout generator (2.2), which now also offers the possibility to integrate martin-brauns shortcuts (H4cker extension) into the layout.

![Legende](https://github.com/8VIM/8VIM/assets/77550697/87421cc2-15f2-4520-9e7f-1d416234fce9)

Those are:

![overview](https://github.com/8VIM/8VIM/assets/77550697/1f7fa7e7-32b6-4949-8ee0-9aa0cf312674)

**Comment by foolsbazar** [comment]
YES, Thank you! now the extra layers make sense!

**Comment by foolsbazar** [comment]
Hi @kjoetom, i was unable to use and then debug your ods document because the generated yaml produces many errors (for instance, it seems that the extra layers should not be more than five).
I created a python script to perform a similar task, but to keep things simple i didn't include all the extra movements you proposed in your file.
If anyone is interested, in attachment they can find the python script and the ods file for a generating the layout. Therefore to generate the yaml perform the following steps:
 * download the zip file and extract the content inside a folder.
 * open the ods file and modify the layout in the first datasheet as desired, without modifying anything inside the red cells. (Notice that when changing the layout via copy-paste action, you should paste-special 'unformatted text', otherwise you'll loose cells references in the second datasheet (no, i'm not going to modify manually each of the 512 cells to avoid that)).
 * then go to the second sheet and modify only those cells where you want to see a different character than when pressing the Shift key (by default letters are automatically capitalized)
 * save the file with the desired name \<filename\>.ods
 * move again to the first sheet save the file as \<filename\>.csv  . Be sure to use the CSV extension so to export properly the layout. and, if prompted, use the following export options: Character set Unicode (UTF-8);  field delimiter , ; string delimiter "; and tick only the options "save cell content as shown"  
 * move again to the second sheet and save the file as \<filename\>-upper.csv with the same options. You can skip this step if you didn't modify anything in the second sheet.
 * open a terminal inside the folder where you've previously extracted the files and run the code 
 `python3 layout_generator.py <filename>.csv`
Other than python3, to work it requires the packages csv, sys, os, but they should come together with python by default.

By the way, only UTF-8 characters are allowed by default.

I tried to consider all the use cases, but probably someone will still find bugs. I hope they'll still find it useful.
[layout_generator.zip](https://github.com/8VIM/8VIM/files/13079131/layout_generator.zip)

**Reply by kjoetom** [reply]
Hi @foolsbazar,

Thanks for your interest and for trying out my layout generator. I am sorry that it did not work for you. I've tested it several times and it worked fine for me. Maybe it has something to do with the fact that I use a German operating system and therefore the formulas in Libreoffice have to be programmed in German. I would have to test this, but unfortunately I do not have the time at the moment and to be honest I do not think this is the reason.

As for the number of layers used, I only used the hidden layer in addition to the first visible one. Using the other extra layers has the advantage that you can see the letters you are going to write in advance, but this is at the expense of having to make a lot of extra movements, which in my opinion disturbs the flow of writing too much.

By the way, I deliberately chose to use a single GUI software to make it possible for people with little computer knowledge to create layouts. On the other hand, I find your combination with Python very exciting, but unfortunately I am not at all familiar with it. In any case, it is good to have an additional option to support trying out new layouts.

**Reply by foolsbazar** [reply]
Yes, I totally understood and appreciated your intentions :-) That's why I tried a bit to work on the office file, but I could not figure it out, probably because I find making complex operations in spreadsheets a bit painful. I didn't understand how  the hidden layer works, so I did not include it, but the extra_layers (which I figured them out only thanks to your file) become very very handy with languages that have many accents for the same letters like French, Italian or Swidish. 

> it is good to have an additional option to support trying out new layouts.

Yes i totally agree :-)

**Comment by kjoetom** [comment]
I have now tested the 8vim-layout-generator with an English version of Libreoffice (using a recent MX Live CD). Everything worked as intended and the resulting yaml file is accepted by 8vim without any problems.

However, I noticed that I had one sentence in the description wrong:  "When you're done, go to the green table "XML result (copy&paste)". This should of course be "YAML result (copy&paste)" and is possibly confusing because I also left the obsolete XML result (a blue sheet) from an earlier version in case someone wants to generate a xml-layout for an older 8vim version.

You did use Libreoffice and not Microsoft Office, right?  :-)

I already have uploaded a corrected version of the layout generator.

kjoe

**Comment by Thain** [comment]
Thanks so much for taking the time to make the spreadsheet, it's an excellent tool! Just a quick note for anyone new, if you're having trouble getting it to work. 

I found the 8vim app just outright crashes if there are issues with your yaml file, which I dealt with at length, and I had to purposefully crash the app a few times before the OS (graphene) would offer to show me the logs (I'm sure there's a better way to do that but I'm not aware). The logs were difficult to read but I eventually figured out what happened:

if, like me, you removed a lot of the "redundant" upper case Greek letters (like A, B) but left the lower case ones (α, β), then the generated yaml will have a bunch of situations like this:
```
...
upper_case: 
lower_case: "α"
...
```
where `upper_case` is left empty. This isn't allowed (at least by 8vim, I don't know yaml) and so when you try to load that keymap file the app crashes. **Basically, make sure every red character on the spreadsheet has an uppercase counterpart (ideally in the same spot).**

Again, thanks for the spreadsheet @kjoetom . In case it interests you, I found that this "bug" also causes some misalignment in the uppercase/lowercase of the red symbols; when an uppercase symbol was missing, sometimes the spreadsheet would take the next one it could find, shifting the uppercase side up by one symbol. Other times, it would just leave the blank. This may give some insight into why the bug is happening, but I'm not sure.

**Reply by kjoetom** [reply]
Thanks, @Thain, for pointing this out. I could reproduce this issue - 8vim crashes when `upper_case` is left empty. It only works again after clearing the app data or storage, not just the cache. I wasn't aware of this, and I'm really sorry if it caused problems for anyone.

I've now updated the spreadsheet to version 2.4, so it inserts `""` if a corresponding letter (`upper_case`/`lower_case`) is empty.

So, instead of:
upper_case: 
lower_case: "α"

we get:
upper_case: ""
lower_case: "α"

This should fix the problem. My own test was successful.
